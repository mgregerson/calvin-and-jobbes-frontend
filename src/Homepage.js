import "./Homepage.css";

/** Homepage
 *
 * Props: user {username, firstName, lastName, email, applications: []}
 * State: None
 *
 * App -> Homepage
 */

function Homepage({ user }) {
  return (
    <div className="Homepage">
      <div className="container text-center my-auto">
        <h1 className="mb-4 fw-bold">Jobber the Hutt</h1>
        {user ? (
          <p className="lead">Welcome, {user.username}!</p>
        ) : (
          <p className="lead">Find Your Dream Job</p>
        )}
      </div>
    </div>
  );
}

export default Homepage;
