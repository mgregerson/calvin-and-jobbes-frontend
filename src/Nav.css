.Navigation {
  background-color: rgb(238, 246, 238);
}

.nav-link {
  color: rgb(99, 91, 91);
  display: block;
  font-size: var(--bs-nav-link-font-size);
  font-weight: bold;
  padding: var(--bs-nav-link-padding-y) var(--bs-nav-link-padding-x);
  text-decoration: none;
  transition: color 0.15s ease-in-out, background-color 0.15s ease-in-out,
    border-color 0.15s ease-in-out;
}

a:link {
  text-decoration: none;
}
