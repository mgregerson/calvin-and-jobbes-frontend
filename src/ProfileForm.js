import React, { useState } from "react";
import { useContext } from "react";
import userContext from "./userContext";

/** ProfileForm
 *
 * Props:
 *       - handleProfileEdit: function passed down from App.
 *
 * State:
 *       - formData: Object with Form Data ({username, firstName, lastName, email})
 *
 * Renders a form so that a current user can edit their profile.
 */

function ProfileForm({ handleProfileEdit }) {
  // USE CONTEXT OF USER: SET CURR FORM DATA TO USERNAME, FIRSTNAME, LASTNAME, EMAIL
  const { user } = useContext(userContext);
  const [formData, setFormData] = useState({
    username: user.username,
    firstName: user.firstName,
    lastName: user.lastName,
    email: user.email,
  });

  console.log(user, "USER IN PROFILEFORM")

  /** Handles keystrokes in searchbar and updates formData */
  function handleChange(evt) {
    const fieldName = evt.target.name;
    const value = evt.target.value;

    setFormData((currData) => {
      currData[fieldName] = value;
      return { ...currData };
    });
  }
  /** handleSubmit of the form. Call function in props */
  function handleSubmit(evt) {
    evt.preventDefault();
    const updateData = formData;
    delete updateData.username;
    console.log(formData, "THE FORM DATA IN PROFILE");
    handleProfileEdit(user.username, updateData);
  }

  return (
    <div className="ProfileForm mb-4 d-flex pt-4">
      <h1 className="ProfileForm-Message">Update Profile</h1>
      <form className="ProfileForm-Form" onSubmit={handleSubmit}>
        <div className="row justify-content-center justify-content-lg-start gx-1">
          <div className="col-8">
            <label htmlFor="username">Username</label>
            <input
              id="username"
              name="username"
              className="form-control form-control-lg"
              onChange={handleChange}
              value={formData.username}
              aria-label="username"
              disabled
            />
            <label htmlFor="firstName">First Name</label>
            <input
              id="first-name"
              name="firstName"
              className="form-control form-control-lg"
              onChange={handleChange}
              value={formData.firstName}
              required
              aria-label="firstName"
            />
            <label htmlFor="lastName">Last Name</label>
            <input
              id="last-name"
              name="lastName"
              className="form-control form-control-lg"
              onChange={handleChange}
              value={formData.lastName}
              aria-label="lastName"
              required
            />
            <label htmlFor="email">Email</label>
            <input
              id="email"
              name="email"
              className="form-control form-control-lg"
              onChange={handleChange}
              value={formData.email}
              aria-label="email"
              required
            />
            <div className="col-auto">
              <button className="btn search-btn btn-lg btn-primary">
                Submit
              </button>
            </div>
          </div>
        </div>
      </form>
    </div>
  );
}

export default ProfileForm;
