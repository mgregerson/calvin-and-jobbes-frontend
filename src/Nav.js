import { NavLink } from "react-router-dom";
import "./Nav.css";
import { useContext } from "react";
import userContext from "./userContext.js";
import jwt_decode from "jwt-decode";

/** Nav
 *
 * Contains links to '/', '/companies', '/jobs', '/login', '/signup', '/profile'
 * and logout functionality when logged in.
 *
 * Contains links to '/login' and '/signup' when not logged in.
 *
 * Props:
 *        - logOut: function that handles logging out a user
 *
 * App -> Nav
 *
 */

function Nav({ logOut }) {
  const { user } = useContext(userContext);
  // token
  return (
    <nav className="Navigation navbar navbar-expand-md">
      <div className="container-fluid">
        <NavLink className="navbar-brand" to="/" end>
          Jobber the Hutt
        </NavLink>

        {user?.username ? (
          <ul className="navbar-nav ms-auto">
            <li className="nav-item me-4">
              <NavLink className="nav-link" to="/companies" end>
                Companies
              </NavLink>
            </li>
            <li className="nav-item me-4">
              <NavLink className="nav-link" to="/jobs" end>
                Jobs
              </NavLink>
            </li>
            <li className="nav-item me-4">
              <NavLink className="nav-link" to="/profile" end>
                Edit Profile
              </NavLink>
            </li>
            <li className="nav-item me-4">
              <NavLink className="nav-link" to="/" onClick={logOut} end>
                Log out {user?.username}
              </NavLink>
            </li>
          </ul>
        ) : (
          <ul className="navbar-nav ms-auto">
            <li className="nav-item me-4">
              <NavLink className="nav-link" to="/login" end>
                Login
              </NavLink>
            </li>
            <li className="nav-item me-4">
              <NavLink className="nav-link" to="/signup" end>
                Sign Up
              </NavLink>
            </li>
          </ul>
        )}
      </div>
    </nav>
  );
}

export default Nav;
