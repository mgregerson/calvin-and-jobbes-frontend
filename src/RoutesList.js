import { Route, Routes, Navigate } from "react-router-dom";
import Homepage from "./Homepage";
import CompanyList from "./CompanyList";
import CompanyDetail from "./CompanyDetail";
import NotFound from "./NotFound";
import JobList from "./JobList";
import LoginForm from "./LoginForm";
import SignupForm from "./SignupForm";
import Unauthorized from "./Unauthorized";
import ProfileForm from "./ProfileForm";
import { useContext } from "react";
import userContext from "./userContext.js";

/** RoutesList
 *
 * Contains site-wide routes + global 404 + global 403 (for non-user accessing protected routes)
 *
 * Props:
 *
 *      - handleSignup: Func
 *      - handleLogin: Func
 *      - handleEditUser: Func
 *
 * App -> RoutesList
 *
 */

function RoutesList({ handleSignup, handleLogin, handleProfileEdit }) {
  const { user } = useContext(userContext);

  console.log(user, "USER IN ROUTES LIST");

  if (user) {
    return (
      <Routes>
        <Route path="/" element={<Homepage user={user} />}></Route>
        <Route path="/companies" element={<CompanyList />}></Route>
        <Route path="/companies/:handle" element={<CompanyDetail />}></Route>
        <Route path="/jobs" element={<JobList />}></Route>
        <Route path="/profile" element={<ProfileForm handleProfileEdit={handleProfileEdit} />}></Route>
        <Route path="*" element={<NotFound />}></Route>
      </Routes>
    );
  }

  return (
    <Routes>
      <Route path="/" element={<Homepage user={user} />}></Route>
      <Route
        path="/login"
        element={<LoginForm handleLogin={handleLogin} />}
      ></Route>
      <Route
        path="/signup"
        element={<SignupForm handleSignup={handleSignup} />}
      ></Route>
      <Route path="/companies" element={<Unauthorized />}></Route>
      <Route path="/companies/:handle" element={<Unauthorized />}></Route>
      <Route path="/jobs" element={<Unauthorized />}></Route>
      <Route path="*" element={<NotFound />}></Route>
    </Routes>
  );
}

export default RoutesList;
