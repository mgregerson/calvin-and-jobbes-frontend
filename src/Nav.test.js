import Nav from "./Nav";
import { render, fireEvent, screen } from "@testing-library/react";
import userContext from "./userContext.js";
import { MemoryRouter } from "react-router-dom";
import RoutesList from "./RoutesList";
import Homepage from "./Homepage";
import ProfileForm from "./ProfileForm";

function testHandleLogout() {
  console.log("Called testHandleLogout");
}

testHandleLogout = jest.fn();

describe("Nav component", function () {
  it("Renders without crashing", function () {
    const user = { username: "test-user" };
    render(
      <userContext.Provider value={{ user }}>
        <MemoryRouter>
          <Nav logOut={testHandleLogout} />
          <Homepage user={user} />
        </MemoryRouter>
      </userContext.Provider>
    );
  });
  it("matches snapshot", function () {
    const user = { username: "test-user" };
    const { container } = render(
      <userContext.Provider value={{ user }}>
        <MemoryRouter>
          <Nav logOut={testHandleLogout} />
          <Homepage user={user} />
        </MemoryRouter>
      </userContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });
  it("Displays the Jobber the Hutt homepage link", function () {
    const user = { username: "test-user" };
    const { container } = render(
      <userContext.Provider value={{ user }}>
        <MemoryRouter>
          <Nav logOut={testHandleLogout} />
        </MemoryRouter>
      </userContext.Provider>
    );
    const homeLink = document.querySelector(".navbar-brand");
    expect(homeLink).toBeInTheDocument();
  });
  it("Directs the user to the homepage when you click the navbar-brand", function () {
    const user = { username: "test-user" };
    render(
      <userContext.Provider value={{ user }}>
        <MemoryRouter>
          <Nav logOut={testHandleLogout} />
          <ProfileForm handleProfileEdit={testHandleLogout} />
        </MemoryRouter>
      </userContext.Provider>
    );
    const navbarBrandLink = screen.getByText("Jobber the Hutt");
    fireEvent.click(navbarBrandLink);

    // Assert that the current URL is "/"
    expect(window.location.pathname).toBe("/");
  });
  // it("Directs the user to the conpanies route when you click Companies", function () {
  //   const user = { username: "test-user" };
  //   render(
  //     <userContext.Provider value={{ user }}>
  //       <MemoryRouter>
  //         <Nav logOut={testHandleLogout} />
  //         <ProfileForm handleProfileEdit={testHandleLogout} />
  //       </MemoryRouter>
  //     </userContext.Provider>
  //   );
  //   const companiesNavLink = screen.getByText("Companies");
  //   fireEvent.click(companiesNavLink);

  //   // // Assert that the current URL is "/companies"
  //   // expect(window.location.pathname).toBe("/companies");
  //   expect(screen.getByText("Company")).toBeInTheDocument();
  // });
});
