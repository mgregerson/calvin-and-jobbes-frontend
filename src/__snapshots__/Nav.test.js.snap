// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav component matches snapshot 1`] = `
<div>
  <nav
    class="Navigation navbar navbar-expand-md"
  >
    <div
      class="container-fluid"
    >
      <a
        aria-current="page"
        class="navbar-brand active"
        href="/"
      >
        Calvin and Jobbes
      </a>
      <ul
        class="navbar-nav ms-auto"
      >
        <li
          class="nav-item me-4"
        >
          <a
            class="nav-link"
            href="/companies"
          >
            Companies
          </a>
        </li>
        <li
          class="nav-item me-4"
        >
          <a
            class="nav-link"
            href="/jobs"
          >
            Jobs
          </a>
        </li>
        <li
          class="nav-item me-4"
        >
          <a
            class="nav-link"
            href="/profile"
          >
            Edit Profile
          </a>
        </li>
        <li
          class="nav-item me-4"
        >
          <a
            class="nav-link"
            href="/applications"
          >
            My Applications
          </a>
        </li>
        <li
          class="nav-item me-4"
        >
          <a
            aria-current="page"
            class="nav-link active"
            href="/"
          >
            Log out 
            test-user
          </a>
        </li>
      </ul>
    </div>
  </nav>
  <div
    class="Homepage"
  >
    <div
      class="container text-center my-auto"
    >
      <h1
        class="Homepage-title mb-4 fw-bold"
      >
        Calvin and Jobbes
      </h1>
      <p
        class="Homepage-welcome lead"
      >
        Welcome, 
        test-user
        !
      </p>
    </div>
  </div>
</div>
`;
